src/Processors/QueryPlan/Optimizations/optimizeJoin.cpp

```diff
@@ -159,247 +322,731 @@ bool optimizeJoinLegacy(QueryPlan::Node & node, QueryPlan::Nodes &, const QueryP
     return true;
 }
 
-void addSortingForMergeJoin(
-    const FullSortingMergeJoin * join_ptr,
-    QueryPlan::Node *& left_node,
-    QueryPlan::Node *& right_node,
+bool convertLogicalJoinToPhysical(
+    QueryPlan::Node & node,
     QueryPlan::Nodes & nodes,
-    const SortingStep::Settings & sort_settings,
-    const JoinSettings & join_settings,
-    const JoinInfo & join_info)
+    const QueryPlanOptimizationSettings & optimization_settings)
+{
+    bool keep_logical = optimization_settings.keep_logical_steps;
+    if (keep_logical)
+        return false;
+    if (!typeid_cast<JoinStepLogical *>(node.step.get()))
+        return false;
+    if (node.children.size() != 2)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "JoinStepLogical should have exactly 2 children, but has {}", node.children.size());
+
+    JoinStepLogical::buildPhysicalJoin(node, optimization_settings, nodes);
+
+    return true;
+}
+
+struct QueryGraphBuilder
 {
-    auto join_kind = join_info.kind;
-    auto join_strictness = join_info.strictness;
-    auto add_sorting = [&] (QueryPlan::Node *& node, const Names & key_names, JoinTableSide join_table_side)
-    {
-        SortDescription sort_description;
-        sort_description.reserve(key_names.size());
-        for (const auto & key_name : key_names)
-            sort_description.emplace_back(key_name);
-
-        auto sorting_step = std::make_unique<SortingStep>(
-            node->step->getOutputHeader(), std::move(sort_description), 0 /*limit*/, sort_settings, true /*is_sorting_for_merge_join*/);
-        sorting_step->setStepDescription(fmt::format("Sort {} before JOIN", join_table_side));
-        node = &nodes.emplace_back(QueryPlan::Node{std::move(sorting_step), {node}});
+    JoinExpressionActions expression_actions;
+
+    std::vector<RelationStats> relation_stats;
+    std::vector<QueryPlan::Node *> inputs;
+
+    std::vector<JoinActionRef> join_edges;
+
+    std::unordered_map<size_t, std::pair<BitSet, JoinKind>> join_kinds;
+    std::unordered_map<size_t, ActionsDAG::NodeRawConstPtrs> type_changes;
```

Comment:
Do we use input number as a key? If so, maybe we can put `join_kinds, type_changes, relation_stats, inputs` into one  `QueryGraphBuilder::Node` structure.